# Exploring Ebay Car Sales Data

# Introduction

The aim of this project is to clean and analyse the data of a car's dataset from _eBay Kleinanzeigen_.

Firstly, we will use Pandas and NumPy libraries for data cleaning, and then we will analyse the included used car listings.

The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/). However, the version of the dataset we are working with is a sample of 50,000 data points that was prepared by [Dataquest](https://www.dataquest.io/).

## Reading the data

We start by importing Pandas and NumPy libraries:

In [1]:
import numpy as np
import pandas as pd

We try different encodings to be able to read the file without error.

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")
#We print some information about the data set
autos.info()
autos.head() #First five rows

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

By this inspection we may notice that our dataset contains 20 columns, most of which are strings. Also, it is important to see that the column names use _camelCase_ instead of Python's preferred _snake_case_, which means we can't just replace spaces with underscores.

Let's start by fixing this problem.

# Data cleaning

In [3]:
#We show the names:
autos.columns
#We change this column to:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
#We check the new dataset:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

### Initial Data Exploration

Now we are going to explore out data using `describe()` to find evident areas where we can continue with our data cleaning.

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

From this table, the following observations can be made:

- The descriptive statistics in **Num_photos** column seems inconsistent.


- **Seller** and **offer_type** columns have nearly all same values, except for one entry.


- **Price** and **odometer** aren't `int()` or `float()` type, but they mean numbers so we need to transform string data to numerical data.


- **Registration_month** has min value of 0 which is not possible as months range in 1 to 12.


- **Registration_year** has min value 1000 which was long before cars were invented and max value 9000 which is many years in the future.


### Num_photos, Seller and offer_type columns.

In [5]:
#Lets analyse this column
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

We are going to delete this colummn as all values are 0. In the same way, we are going to drop seller and offer_type columns for having same values.

In [6]:
autos = autos.drop(["seller","offer_type","num_photos"], axis = 1)
autos.shape

(50000, 17)

### Price and odometer columns.

We need to convert these columns to numerical data type columns as they are meant to be numbers.

In [7]:
#We print what we have to change
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [8]:
#We change price column
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','').astype(int)

In [9]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [10]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','').astype(int)
autos = autos.rename({'odometer':'odometer_km'},axis = 1)
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0       5000  control          bus               2004    manuell       158   
1       8500  control    limousine               1997  automatik       286   
2       8990     test    limousine               2009    manuell       102   
3       4350  control   kleinwagen               2007  automatik        71   
4       1350     test        kombi               2003    manuell         0   
...      ...      ...          ...                ...        ...       ...   
49995  24900  control    limousine               2011  automatik       239   
49996   1980  control       cabrio               1996    manuell        75   
49997  13200     test       cabrio               2014  automatik        69   
49998  22900  control        kombi               2013    manuell       150   
49999   1250  control    limousine               1996    manuell       101   

        model  odometer_km  registration_month fuel_type       brand  \
0      andere       150000                   3       lpg     peugeot   
1         7er       150000                   6    benzin         bmw   
2        golf        70000                   7    benzin  volkswagen   
3      fortwo        70000                   6    benzin       smart   
4       focus       150000                   7    benzin        ford   
...       ...          ...                 ...       ...         ...   
49995      q5       100000                   1    diesel        audi   
49996   astra       150000                   5    benzin        opel   
49997     500         5000                  11    benzin        fiat   
49998      a3        40000                  11    diesel        audi   
49999  vectra       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[50000 rows x 17 columns]

Now we can analyse these two columns to see if there are other changes that are needed to be made.

In [11]:
print(autos['price'].unique().shape)
autos['price'].describe()

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [12]:
autos['price'].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

Some price values are abnormally low and others high. Some are as low as 0 and some even reach 100 million. We will remove these outliers before we investigate further. We will set the minimum price to be 1000 and maximum as 100000 dollars.

In [13]:
autos.loc[autos['price'].between(0,1000),'price'] = np.nan
autos.loc[autos['price'].between(100000,100000000),'price'] = np.nan
autos['price'].value_counts().sort_index()

1039.0      1
1040.0      1
1049.0      6
1050.0     95
1059.0      1
           ..
93911.0     1
94999.0     1
98500.0     1
99000.0     2
99900.0     2
Name: price, Length: 2051, dtype: int64

In [14]:
#Let's see how the statistics have changed
print(autos['price'].value_counts().shape)
autos['price'].describe()

(2051,)


count    37948.000000
mean      7191.577896
std       7920.021774
min       1039.000000
25%       2300.000000
50%       4499.000000
75%       8999.000000
max      99900.000000
Name: price, dtype: float64

Now that we have removed the outliers, we are left with around 38000 price entries. Standard deviation has also been reduced from about 480000 to almost 8000 which is a direct result of removing the extremely high and small price values.

Let's analyse Odometer column:

In [15]:
print(autos['odometer_km'].unique().shape)
autos["odometer_km"].describe()

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [16]:
autos["odometer_km"].value_counts().sort_index()

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

The data in this column seems fine and it does not have any outliers that could affect our analysis. We can easily observe that there are more high mileage than low mileage vehicles.

### Date columns.

As we saw earlier, there were a few invalid listings of year and month.

For analysing the problem properly, first we are going to look at the apparently correct date columns.

In [17]:
#Let's print just the first five rows
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [21]:
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

We observe that all the data, corresponding to this column, is from March and April of 2016. This makes us think that the website was probably crawled around that time. Furthermore, we can see that all values move in a reasonable range, so this column is not going to be dropped.

In [23]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In this case, we see that the range of dates are spread over 10 months. However, most fall within March and April 2016. This makes sense, as in these months the website was disproportionately crawled.

In [25]:
(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

We might state that the crawler was collecting data from the site between March and April 2016. This is clearly seen in both _date_crawled_ and _last_seen_ columns. 

However, in this column, there is a remarkable discrepancy in the last three days, as these values are 6-10x higher than the previous days. The _last_seen_ column recorded the date it last saw any listing. This could mean that 6-10x more sales were made on the last three days above. This massive spike in sales is odd and highly unlikely, therefore, we can conclude that the above spike had more to do with the ending of the crawling period rather than with the sales.


### Incorrect data in date columns

Once we've analysed some date columns, we can say that any vehicle with a registration year above 2016 will be definitely inaccurate, due to the fact that a car can't be first registered before the listing was seen. After some googling I found that the Germany started registering number plates in 1896. So we can have 1896 as the lowest value.



In [27]:
autos=autos[autos['registration_year'].between(1896,2016)]
autos['registration_year'].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
          ...   
2012    0.027546
2013    0.016782
2014    0.013867
2015    0.008308
2016    0.027401
Name: registration_year, Length: 78, dtype: float64

In [28]:
autos['registration_year'].describe()

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64

We can see that almost 75% of the cars were registered after 1999 and amongst these 50% were registered in the early 2000s.

Now let's look at the registration_month column. Before we saw that it had a min value of 0, which it's not possible.

In [29]:
autos["registration_month"].value_counts()

3     4899
0     4587
6     4211
4     3954
5     3949
7     3812
10    3554
12    3325
9     3301
11    3271
1     3164
8     3085
2     2916
Name: registration_month, dtype: int64

Around 4500 entries are from the month 0, which makes no sense. Since there is no way we can find the actual month, it is most suitable that we remove these listings.

In [30]:
autos = autos[~(autos["registration_month"]==0)]
autos["registration_month"].value_counts()

3     4899
6     4211
4     3954
5     3949
7     3812
10    3554
12    3325
9     3301
11    3271
1     3164
8     3085
2     2916
Name: registration_month, dtype: int64

# Data analysis

Once we've cleaned our data we can proceed with data analysis.

### 1-  Analysing Price By Brand

Let's start by seeing how popular each brand is.

In [32]:
autos['brand'].value_counts(normalize=True)*100

volkswagen        20.860477
bmw               11.219815
opel              10.446353
mercedes_benz      9.919201
audi               8.798140
ford               6.848369
renault            4.684515
peugeot            2.981055
fiat               2.548284
seat               1.832370
skoda              1.680440
nissan             1.537718
mazda              1.493980
smart              1.427223
citroen            1.381184
toyota             1.319030
hyundai            1.012868
sonstige_autos     0.959923
volvo              0.929997
mini               0.925393
mitsubishi         0.794181
honda              0.784973
kia                0.755047
alfa_romeo         0.656062
porsche            0.637646
suzuki             0.600815
chevrolet          0.570889
chrysler           0.363712
dacia              0.271633
daihatsu           0.239405
jeep               0.232499
land_rover         0.214083
subaru             0.204876
saab               0.168044
jaguar             0.161138
daewoo             0

We can see that more than 50% of the cars are German, with Volkswagon being on the top. Japanese brands occupy 20-25% of the market, and the rest of the market is occuppied by American, Korean, French, and Italian.

Volkswagen is the most popular brand. Note that a lot of brands do not even account for 1% of the listings. Since they do not affect our analysis much, we will limit our analysis to brands that account for at least 1% of our total listings. Our aim is to find out the mean prices of popular brands.

Let's see which is the most expensive.

In [33]:
#We delete non popular brands
brand_perc = autos["brand"].value_counts(normalize=True)
popular = brand_perc[brand_perc>0.01].index

#We show the price of the popular brands 
brand_mean_prices = {}
for brand in popular:
    rows = autos[autos["brand"]==brand]
    mean_price = rows["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 6935,
 'bmw': 9273,
 'opel': 4378,
 'mercedes_benz': 9432,
 'audi': 10638,
 'ford': 5452,
 'renault': 3736,
 'peugeot': 4108,
 'fiat': 4165,
 'seat': 5761,
 'skoda': 7008,
 'nissan': 6572,
 'mazda': 5566,
 'smart': 3895,
 'citroen': 4709,
 'toyota': 5668,
 'hyundai': 6356}

According to the above analysis:

- Audi is the most expensive car and Renault is the least expensive car in our selected top German brands.

- Volkswagen being the most common German brand has a mean price of 6935 which is 3700 less than Audi.

Comparing this result to the popularity, Volkswagen is the most popular probably due to its performance over its price. However, just this much is not enough to conclude that mid range cars are popular in general

### 2- Analysing Mileage by brand

In [36]:
#We show brands and milleage

brand_mean_mileage = {}
for brand in popular:
    rows = autos[autos["brand"]==brand]
    mean_mileage = rows["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

#We print means
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)
popularity = pd.Series(autos["brand"].value_counts(normalize=True).sort_values(ascending=False))

In [38]:
brand_info = pd.DataFrame(mean_mileage,columns=["mean_mileage"])
brand_info["popularity"] = popularity
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  popularity  mean_price
bmw                  132541    0.112198        9273
mercedes_benz        130971    0.099192        9432
opel                 128884    0.104464        4378
audi                 128784    0.087981       10638
volkswagen           128343    0.208605        6935
renault              127872    0.046845        3736
peugeot              126216    0.029811        4108
ford                 123983    0.068484        5452
mazda                123543    0.014940        5566
seat                 121017    0.018324        5761
citroen              118516    0.013812        4709
nissan               117485    0.015377        6572
fiat                 116038    0.025483        4165
toyota               115191    0.013190        5668
skoda                110157    0.016804        7008
hyundai              105613    0.010129        6356
smart                 99104    0.014272        3895

We can see that, in general, there is a slight trend to the most expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage. However, there are exceptions, for example, Mercedez Benz has a higher mean price and a lower mean mileage than BMW.

### 3- Price difference between Repaired and Non repaired cars

In [40]:
autos.groupby('unrepaired_damage')['price'].mean()

unrepaired_damage
ja      4172.917485
nein    8025.298255
Name: price, dtype: float64

Non-damaged cars cost approximately $3850 more than damaged cars.

### 4- Verifying the trend between Odometer & Price

In [42]:
print(autos['odometer_km'].value_counts().sort_values(ascending=False))

150000    27646
125000     4659
100000     1949
90000      1612
80000      1336
70000      1147
60000      1096
50000       966
40000       779
30000       740
20000       722
5000        564
10000       225
Name: odometer_km, dtype: int64


In [44]:
autos.groupby(by='odometer_km')['price'].mean()

odometer_km
5000      12086.057534
10000     21829.806931
20000     18415.302432
30000     16502.218130
40000     15691.964427
50000     14129.940043
60000     12866.201149
70000     11614.660878
80000     10372.839172
90000      9501.547144
100000     9078.391304
125000     7221.331820
150000     5038.730090
Name: price, dtype: float64

Price and odometer have an inverse relationship as seen above. Price starts to strictly follow this correlation at 10,000km group where it is 21,829 and then it decreases to 5,038 as odometer moves to 150,000km.

# Conclusion

To put the matter in a nutshell, in this analysis, it was found that more than 50% of the market is covered by German car manufacturers. Amongst these brands, Volkswagen has the highest number of cars with a decent mean price. Furthermore, a really interesting inverse correlation between the mean price and mean milleage, amongst the top brands ,was also discovered.